File: src/api/movies.ts
import { apiClient } from "./client";
import type { Movie, MovieListResponse } from "../types/movie";

export type MovieListKind = "popular" | "top-rated" | "now-playing" | "upcoming";

export async function getMovieList(kind: MovieListKind, page = 1) {
    const res = await apiClient.get<MovieListResponse>(`/api/movies/${kind}`, {
        params: { page },
    });
    return res.data;
}

export async function searchMovies(query: string, page = 1) {
    const res = await apiClient.get<MovieListResponse>("/api/movies/search", {
        params: { query, page },
    });
    return res.data;
}

export async function getMovie(movieId: number) {
    const res = await apiClient.get<Movie>(`/api/movies/${movieId}`);
    return res.data;
}

File: backend/app/routers/movies.py
from fastapi import APIRouter, Query

from app.schemas.movie import Movie, MovieListResponse
from app.services.tmdb_service import tmdb_service


router = APIRouter(prefix="/movies", tags=["movies"])


@router.get("/popular", response_model=MovieListResponse)
async def popular_movies(page: int = Query(default=1, ge=1)) -> MovieListResponse:
    return await tmdb_service.get_popular(page=page)


@router.get("/top-rated", response_model=MovieListResponse)
async def top_rated_movies(page: int = Query(default=1, ge=1)) -> MovieListResponse:
    return await tmdb_service.get_top_rated(page=page)


@router.get("/now-playing", response_model=MovieListResponse)
async def now_playing_movies(page: int = Query(default=1, ge=1)) -> MovieListResponse:
    return await tmdb_service.get_now_playing(page=page)


@router.get("/upcoming", response_model=MovieListResponse)
async def upcoming_movies(page: int = Query(default=1, ge=1)) -> MovieListResponse:
    return await tmdb_service.get_upcoming(page=page)


@router.get("/search", response_model=MovieListResponse)
async def search_movies(
    query: str = Query(min_length=1),
    page: int = Query(default=1, ge=1),
) -> MovieListResponse:
    return await tmdb_service.search(query=query, page=page)


@router.get("/{movie_id}", response_model=Movie)
async def movie_detail(movie_id: int) -> Movie:
    return await tmdb_service.get_movie(movie_id=movie_id)
